DEVX-1263 DEVX-1269 Docs Conversion: secret protection beginner cloud (#767)
confluentinc/examples: security/secret-protection/docs/secret-protection.rst

.. _secret-protection-tutorial:

> Tuturial: Secret Protection

Reviewer: wrt TOC, what do you think about adding it under https://docs.confluent.io/current/security/secrets.html (requires PR to docs)

Author: Yes I think that's a good place.

Reviewer: Cool.  Can you create a PR to docs to add it there?

Author: Yes will do.

===========================

Overview
--------

> Confluent’s **Secret Protection** feature encrypts secrets in configuration

Reviewer: I don't think `Secret Protection` should be bold (here or in the line below)?

Author: You're right. Will update

files. Instead of storing passwords or other sensitive data as cleartext,
Secret Protection encrypts the data within a configuration file itself.

.. note::

>      You can either step through this tutorial, or run the `automated

Reviewer: This link won't work without the `:devx-examples:` type of tag

Author: Updated

     demo <demo-secret-protection.sh>`__.

Prerequisites
~~~~~~~~~~~~~

> #. Download `Confluent Platform 5.4 <https://www.confluent.io/download/>`__ or greater

Reviewer: Do not pin CP version.  Use same syntax used in clients example (with just link to download)

Author: Got it - updated.

> #. Get the new `Confluent

Reviewer: You can remove this step because CLI is back to being bundled with CP

Author: Updated

   CLI <https://docs.confluent.io/current/cli/installing.html>`__ (v0.128.0 or
   greater)

Workflow
--------

In the most common use case, you would want to encrypt passwords. Confluent's
`Security tutorial
<https://docs.confluent.io/current/tutorials/security_tutorial.html>`__ shows

> you how to enable security features on the |cp|, but requires you to take extra

Reviewer: ```suggestion
you how to enable security features on the |cp|, but that requires you to take extra
```

Author: Updated

> steps in generating keys and certificates, and adding TLS configurations. In

Reviewer: FYC: `Instead in this tutorial...`

> this tutorial instead, you'll encrypt a basic configuration parameter, which are

Reviewer: Let's remove `which are the same steps`

the same steps.

Generating the master encryption key based on a passphrase
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> #. Select a master encryption key passphrase.

Reviewer: ```suggestion
#. Decide on a master encryption key passphrase.
```

>    .. tip::

Reviewer: Too many tips are distracting.  Can we integrate the statement in the tip up into the step above?

        Your passphrase should be longer than the usual password. Choose a
        phrase you can remember as a string of words.

> #. Enter the passphrase into a file to be passed into the CLI, to avoid logging

Reviewer: ```suggestion
#. Write the passphrase in a local file (e.g. ``/path/to/passphrase.txt``), which will be passed into the CLI, to avoid logging
```

   history showing the passphrase.

#. Choose the location where the secrets file will reside on your local host.

>    .. note::

Reviewer: Too many notes are distracting. Can we integrate the statement in the note up into the step above?

       The location shouldn't be where the |cp| services run.

>    The secrets file will contain encrypted secrets for the master encryption

Reviewer: These 3 lines should be added to the workflow description.  It's not a step

   key, data encryption key, and configuration parameters, along with their
   metadata, such as which cipher was used for encryption.

#. Generate the master encryption key by running the following command:

   .. code-block:: text

      # passphrase: /path/to/passphrase.txt
      # local-secrets-file: /path/to/secrets.txt
      confluent secret master-key generate --local-secrets-file /path/to/secrets.txt --passphrase @/path/to/passphrase.txt

   You should see:

   .. code-block:: text

      Save the master key. It cannot be retrieved later.
      +------------+----------------------------------------------+
      | Master Key | Nf1IL2bmqRdEz2DO//gX2C+4PjF5j8hGXYSu9Na9bao= |
      +------------+----------------------------------------------+

#. Save the master key somewhere safe on your computer.

> #. Export the key into the environment on the local host and every host

Reviewer: ```suggestion
#. Export the key into the environment on the local host and every other host
```

>    that will have a configuration file with secret protection by running the

Reviewer: ```suggestion
   that will have a configuration file with secret protection, by running the
```

   following command:

   .. code-block:: text

      export CONFLUENT_SECURITY_MASTER_KEY=Nf1IL2bmqRdEz2DO//gX2C+4PjF5j8hGXYSu9Na9bao=

>    To protect the previous environment variable in a production host, you can set

Reviewer: ```suggestion
#. (Optional) To protect the previous environment variable in a production host, you can set
```

   the master encryption key at the process level instead of the global machine
   level. For example, you could set it in the ``systemd`` overrides for executed
   processes, restricting the environment directives file to root-only access.

Encrypting the value of a configuration parameter
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Use a configuration parameter available in the configuration file example that

Reviewer: ```suggestion
In this section, you will encrypt a configuration parameter available in the configuration file example that
```

ships with |cp|.

To encrypt the parameter ``config.storage.topic`` in
``$CONFLUENT_HOME/etc/schema-registry/connect-avro-distributed.properties``,
complete the following steps:

> #. Make a backup of this file, because the CLI currently does in-place

Reviewer: ```suggestion
#. Make a backup of the file ``$CONFLUENT_HOME/etc/schema-registry/connect-avro-distributed.properties``, because the CLI currently does in-place
```

   modification on the original file.

#. Select the exact path where the secrets file will reside on the remote hosts

>    where the |cp| services run.

Reviewer: ```suggestion
   where the |cp| services run, e.g. ``/path/to/secrets-remote.txt``
```

#. Encrypt the field:

   .. code-block:: text

         # remote-secrets-file: /path/to/secrets-remote.txt
         confluent secret file encrypt --local-secrets-file /path/to/secrets.txt --remote-secrets-file /path/to/secrets-remote.txt --config-file connect-avro-distributed.properties --config config.storage.topic

   .. code-block:: text

>          # Value before encryption

Reviewer: The order may be misleading for users.  Can we do the following?  First tell users to look at the value before encryption.  Then encrypt it (with command above), then tell users to look at the value after encryption

   .. code-block:: text

      ...
      connect-avro-distributed.properties/config.storage.topic = ENC[AES/CBC/PKCS5Padding,data:CUpHh5lRDfIfqaL49V3iGw==,iv:vPBmPkctA+yYGVQuOFmQJw==,type:str]

Decrypting the value of a configuration parameter
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can also decrypt the value of a configuration parameter into a file by
completing the following steps:

#. Run the following command to decrypt the value into a file:

   .. code-block:: bash

      confluent secret file decrypt --local-secrets-file /path/to/secrets.txt --config-file connect-avro-distributed.properties --output-file decrypted.txt

#. View the file:

   .. code-block:: bash

      cat decrypted.txt

   You should see the following output:

   .. code-block:: bash

        config.storage.topic = connect-configs

Updating the value of the configuration parameter
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You may need to update secrets on a regular basis to ensure the secrets don't
get stale. The configuration parameter ``config.storage.topic`` was originally
set to ``connect-configs``. If you must change this value in the future, you can
update it using the CLI.

> In the following CLI, pass in a file ``/path/to/updated-config-and-value`` that

Reviewer: This and following steps should be numbered with `#.`

The configuration file ``connect-avro-distributed.properties`` doesn't change
because it’s a pointer to the secrets file, but the secrets file has a new value
for the encrypted value for this file/parameter pair.

When running the following command:

.. code-block:: bash

   cat /path/to/secrets.txt

You should see:

.. code-block:: bash

   ...
   connect-avro-distributed.properties/config.storage.topic = ENC[AES/CBC/PKCS5Padding,data:CblF3k1ieNkFJzlJ51qAAA==,iv:dnZwEAm1rpLyf48pvy/T6w==,type:str]

Trust but verify
~~~~~~~~~~~~~~~~

That’s cool! But does it work? Try it out yourself. Run |ak| and start the
modified connect worker with the encrypted value of
``config.storage.topic=newTopicName`` by completing the following steps:

#. Start |zk| and a |ak| broker:

   .. code-block:: bash

      confluent local start kafka

#. Run the modified |kconnect| worker:

   .. code-block:: bash

      connect-distributed connect-avro-distributed.properties > connect.stdout 2>&1 &

> #. List the topics:

Reviewer: ```suggestion
#. List the topics and notice ``newTopicName``
```
